File: c-sharp/Math/src/Tuple.cs
namespace Raytracer.Math
{
    public class Tuple
    {
        public float x { get; set; }
        public float y { get; set; }
        public float z { get; set; }

        // Might need to rethink this, but this is how the book
        // writes these. Let's get it written then iterate.
        //
        // W == 1 - point
        // W == 0 - vector
        public float w { get; set; }

        public Tuple(float x, float y, float z, float w)
        {
            this.x = x;
            this.y = y;
            this.z = z;
            this.w = w;
        }

        public static Tuple operator+(Tuple a, Tuple b)
        {
            return new Tuple(
                a.x + b.x,
                a.y + b.y,
                a.z + b.z,
                a.w + b.w
            );
        }

        public static Tuple operator-(Tuple a)
        {
            return new Tuple(
                0 - a.x,
                0 - a.y,
                0 - a.z,
                0 - a.w
            );
        }

        public static Tuple operator*(Tuple a, double factor)
        {
            return new Tuple(
                a.x * (float)factor,
                a.y * (float)factor,
                a.z * (float)factor,
                a.w * (float)factor
            );
        }

        public static Tuple operator/(Tuple a, double factor)
        {
            return new Tuple(
                a.x / (float)factor,
                a.y / (float)factor,
                a.z / (float)factor,
                a.w / (float)factor
            );
        }
    }
}

File: cpp/the-ray-tracer-challenge/src/tuple.cpp
#include "tuple.h"

tuple::tuple(float x, float y, float z, float w) : _e{ x, y, z, w }
{
};

tuple::tuple(const tuple& t) : tuple(t.x(), t.y(), t.z(), t.w())
{
	std::cout << "tuple copy" << std::endl;
};

float tuple::x() const { return _e[0]; };
float tuple::y() const { return _e[1]; };
float tuple::z() const { return _e[2]; };
float tuple::w() const { return _e[3]; };

tuple tuple::operator+(const tuple& rhs) const
{
	return tuple{
		x() + rhs.x(),
		y() + rhs.y(),
		z() + rhs.z(),
		w() + rhs.w()
	};
}

bool tuple::operator==(const tuple& rhs) const
{
	return floatEqual(x(), rhs.x())
		&& floatEqual(y(), rhs.y())
		&& floatEqual(z(), rhs.z())
		&& floatEqual(w(), rhs.w());
}

tuple tuple::operator-(const tuple& rhs) const
{
	return tuple{
		x() - rhs.x(),
		y() - rhs.y(),
		z() - rhs.z(),
		w() - rhs.w()
	};
}

tuple tuple::operator-() const
{
	return tuple{
		-x(),
		-y(),
		-z(),
		-w()
	};
}

tuple tuple::operator*(float rhs) const
{
	return tuple{
		x() * rhs,
		y() * rhs,
		z() * rhs,
		w() * rhs
	};
}

tuple tuple::operator/(float rhs) const
{
	return tuple{
		x() / rhs,
		y() / rhs,
		z() / rhs,
		w() / rhs
	};
}